lgc/patch/VertexFetch.cpp

```diff
@@ -25,28 +25,140 @@
+// Pass to lower vertex fetch calls
+class LowerVertexFetch : public ModulePass {
+public:
+  LowerVertexFetch() : ModulePass(ID) {}
+  LowerVertexFetch(const LowerVertexFetch &) = delete;
+  LowerVertexFetch &operator=(const LowerVertexFetch &) = delete;
+
+  void getAnalysisUsage(AnalysisUsage &analysisUsage) const override {
+    analysisUsage.addRequired<PipelineStateWrapper>();
+  }
+
+  virtual bool runOnModule(Module &module) override;
+
+  static char ID; // ID of this pass
+};
+
+// =====================================================================================================================
+// Vertex fetch manager
+class VertexFetchImpl : public VertexFetch {
+public:
+  VertexFetchImpl(LgcContext *lgcContext);
+  VertexFetchImpl(const VertexFetchImpl &) = delete;
+  VertexFetchImpl &operator=(const VertexFetchImpl &) = delete;
+
+  // Generate code to fetch a vertex value
+  Value *fetchVertex(Type *inputTy, const VertexInputDescription *description, unsigned location, unsigned compIdx,
+                     BuilderBase &builder) override;
+
+private:
+  void initialize(PipelineState *pipelineState);
+
+  static VertexFormatInfo getVertexFormatInfo(const VertexInputDescription *description);
+
+  // Gets variable corresponding to vertex index
+  Value *getVertexIndex() { return m_vertexIndex; }
+
+  // Gets variable corresponding to instance index
+  Value *getInstanceIndex() { return m_instanceIndex; }
+
+  static const VertexCompFormatInfo *getVertexComponentFormatInfo(unsigned dfmt);
+
+  unsigned mapVertexFormat(unsigned dfmt, unsigned nfmt) const;
+
+  Value *loadVertexBufferDescriptor(unsigned binding, BuilderBase &builder);
+
+  void addVertexFetchInst(Value *vbDesc, unsigned numChannels, bool is16bitFetch, Value *vbIndex, unsigned offset,
+                          unsigned stride, unsigned dfmt, unsigned nfmt, Instruction *insertPos, Value **ppFetch) const;
+
+  bool needPostShuffle(const VertexInputDescription *inputDesc, std::vector<Constant *> &shuffleMask) const;
+
+  bool needPatchA2S(const VertexInputDescription *inputDesc) const;
+
+  bool needSecondVertexFetch(const VertexInputDescription *inputDesc) const;
+
+  LgcContext *m_lgcContext = nullptr;   // LGC context
+  LLVMContext *m_context = nullptr;     // LLVM context
+  Value *m_vertexBufTablePtr = nullptr; // Vertex buffer table pointer
+  Value *m_vertexIndex = nullptr;       // Vertex index
+  Value *m_instanceIndex = nullptr;     // Instance index
+
+  static const VertexCompFormatInfo MVertexCompFormatInfo[]; // Info table of vertex component format
+  static const BufFormat MVertexFormatMap[];                 // Info table of vertex format mapping
+
+  // Default values for vertex fetch (<4 x i32> or <8 x i32>)
+  struct {
+    Constant *int8;     // < 0, 0, 0, 1 >
+    Constant *int16;    // < 0, 0, 0, 1 >
+    Constant *int32;    // < 0, 0, 0, 1 >
+    Constant *int64;    // < 0, 0, 0, 0, 0, 0, 0, 1 >
```

Comment:
Shouldn't this be `< 0, 0, 0, 0, 0, 0, 1, 0 >`? Though I suppose that's not your bug...

Reply:
Yes, should be what you have said. It is my bug :)